src/shell/shell_cmds.cpp

```diff
@@ -465,26 +472,250 @@ static std::string to_search_pattern(const char *arg)
+			Bit16u date = iter->date;
+			Bit16u time = iter->time;
+			Bit8u attr = iter->attr;
+
+			/* output the file */
+			if (optB) {
+				// this overrides pretty much everything
+				if (strcmp(".",uselfn&&!optZ?lname:name) && strcmp("..",uselfn&&!optZ?lname:name)) {
+					shell->WriteOut("%s\n",uselfn&&!optZ?lname:name);
+				} else {
+					// skip to the next file, otherwise this file
+					// will be counted as printed for pause cmd
+					continue;
+				}
+			} else {
+				if (first&&optS) {
+					first=false;
+					shell->WriteOut("\n");
+					shell->WriteOut(MSG_Get("SHELL_CMD_DIR_INTRO"),uselfn&&!optZ&&DOS_GetSFNPath(path,largs,true)?largs:sargs);
+					if (optP) {
+						p_count+=optW?10:2;
+						if (p_count%(GetPauseCount()*w_size)<3) {
+							shell->WriteOut(MSG_Get("SHELL_CMD_PAUSE"));
+							Bit8u c;Bit16u n=1;
+							DOS_ReadFile(STDIN,&c,&n);
+							if (c==3) {shell->WriteOut("^C\r\n");return false;}
+							if (c==0) DOS_ReadFile(STDIN,&c,&n); // read extended key
+							shell->WriteOut_NoParsing("\n");
+						}
+					}
+				}
+				char * ext = empty_string;
+				if (!optW && (name[0] != '.')) {
+					ext = strrchr(name, '.');
+					if (!ext) ext = empty_string;
+					else *ext++ = 0;
+				}
+				Bit8u day	= (Bit8u)(date & 0x001f);
+				Bit8u month	= (Bit8u)((date >> 5) & 0x000f);
+				Bit16u year = (Bit16u)((date >> 9) + 1980);
+				Bit8u hour	= (Bit8u)((time >> 5 ) >> 6);
+				Bit8u minute = (Bit8u)((time >> 5) & 0x003f);
+
+				if (attr & DOS_ATTR_DIRECTORY) {
+					if (optW) {
+						shell->WriteOut("[%s]",name);
+						size_t namelen = strlen(name);
+						if (namelen <= 14) {
+							for (size_t i=14-namelen;i>0;i--) shell->WriteOut(" ");
+						}
+					} else {
+						shell->WriteOut("%-8s %-3s   %-16s %02d-%02d-%04d %2d:%02d %s\n",name,ext,"<DIR>",day,month,year,hour,minute,uselfn&&!optZ?lname:"");
+					}
+					dir_count++;
+				} else {
+					if (optW) {
+						shell->WriteOut("%-16s",name);
+					} else {
+						FormatNumber(size,numformat);
+						shell->WriteOut("%-8s %-3s   %16s %02d-%02d-%04d %2d:%02d %s\n",name,ext,numformat,day,month,year,hour,minute,uselfn&&!optZ?lname:"");
+					}
+					if (optS) {
+						cfile_count++;
+						cbyte_count+=size;
+					}
+					file_count++;
+					byte_count+=size;
+				}
+				if (optW) w_count++;
+			}
+			if (optP && !(++p_count%(GetPauseCount()*w_size))) {
+				if (optW&&w_count%5) {shell->WriteOut("\n");w_count=0;}
+				shell->WriteOut(MSG_Get("SHELL_CMD_PAUSE"));
+				Bit8u c;Bit16u n=1;
+				DOS_ReadFile(STDIN,&c,&n);
+				if (c==3) {shell->WriteOut("^C\r\n");return false;}
+				if (c==0) DOS_ReadFile(STDIN,&c,&n); // read extended key
+				shell->WriteOut_NoParsing("\n");
+			}
+		}
+
+		if (!results.size())
+			found=false;
+		else if (optW&&w_count%5)
+			shell->WriteOut("\n");
+	} else
+		found=false;
+	if (!found&&!optB&&!optS) {
+		shell->WriteOut(MSG_Get("SHELL_CMD_FILE_NOT_FOUND"),args);
+		if (!dirPaused(shell, w_size, optP, optW)) return false;
+	}
```

Comment:
At this point into this function.. it's a gigantic nested web of branches.
I suspect PVS might flag it in cyclomatic complexity. 
Suggest disolving all this down into functions - and get rid of so many nested branches.
This effort will likely change the code (and might introduce bugs), at the same time, it might also reveal existing bugs in the branches.  So suggest some amount of testing where all the branches can be exercised.

Reply:
This part uses the exactly same working code as in DOSBox-X, so I think the code should work as is in spite of its complexity as you said. It is possible to dissolve this to sub-functions, but it indeed likely changes the code and might introduce bugs. So perhaps a cleanup PR to do this later?